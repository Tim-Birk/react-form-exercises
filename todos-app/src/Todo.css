.Todo {
  display: flex;
  margin-bottom: 5px;
}

.Todo-task {
  margin-right: 5px;
}
