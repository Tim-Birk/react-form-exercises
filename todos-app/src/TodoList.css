.TodoList {
  padding-left: 0;
  margin-top: 10px;
}
