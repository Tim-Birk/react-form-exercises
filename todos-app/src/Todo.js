import './Todo.css';

const Todo = ({ id, task, removeTodo }) => {
  const handleRemove = (evt) => {
    removeTodo(id);
  };

  const testId = task ? task.replace(/[^A-Z0-9]/gi, '_') : '';
  return (
    <li data-testid={`Todo-${testId}`} className='Todo'>
      <div className='Todo-task'>- {task}</div>
      <button data-testid={`Todo-delete-${testId}`} onClick={handleRemove}>
        x
      </button>
    </li>
  );
};

export default Todo;
