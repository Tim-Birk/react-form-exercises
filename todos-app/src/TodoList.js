import React, { useState } from 'react';
import Todo from './Todo';
import NewTodoForm from './NewTodoForm';
import { v4 as uuid } from 'uuid';
import './TodoList.css';

const TodoList = () => {
  const [Todos, setTodos] = useState([
    { id: uuid(), task: 'Walk the dog' },
    { id: uuid(), task: 'Make breakfast' },
    { id: uuid(), task: 'Study coding' },
    { id: uuid(), task: 'Exercise' },
  ]);

  const addTodo = (task) => {
    const newTodo = { id: uuid(), task };
    setTodos((oldTodos) => [...oldTodos, newTodo]);
  };

  const removeTodo = (id) => {
    setTodos(Todos.filter((Todo) => Todo.id !== id));
  };

  return (
    <>
      <h1>Todo List</h1>
      <NewTodoForm addTodo={addTodo} />
      <ul className='TodoList'>
        {Todos.map((t) => (
          <Todo key={t.id} id={t.id} task={t.task} removeTodo={removeTodo} />
        ))}
      </ul>
    </>
  );
};

export default TodoList;
