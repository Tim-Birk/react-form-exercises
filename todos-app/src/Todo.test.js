import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Todo from './Todo';
import TodoList from './TodoList';

it('renders without crashing', function () {
  render(<Todo />);
});

it('matches snapshot', function () {
  const { asFragment } = render(<Todo task='This is a test' />);
  expect(asFragment()).toMatchSnapshot();
});

it('can remove a new Todo', function () {
  const { getByLabelText, queryByText, queryByTestId } = render(<TodoList />);

  // no Todo yet
  expect(queryByText('- This is a test')).not.toBeInTheDocument();

  const taskInput = getByLabelText('Task:');
  const submitBtn = queryByText('Add New');

  // fill out the form
  fireEvent.change(taskInput, { target: { value: 'This is a test' } });
  fireEvent.click(submitBtn);

  // Todo exists
  expect(queryByText('- This is a test')).toBeInTheDocument();
  const TodoDeleteBtn = queryByTestId('Todo-delete-This_is_a_test');

  // Todo delete button exists
  expect(TodoDeleteBtn).toBeInTheDocument();
  fireEvent.click(TodoDeleteBtn);
  // Todo removed
  expect(queryByText('- This is a test')).not.toBeInTheDocument();
});
