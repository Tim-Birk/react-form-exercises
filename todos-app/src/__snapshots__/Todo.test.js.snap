// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <li
    class="Todo"
    data-testid="Todo-This_is_a_test"
  >
    <div
      class="Todo-task"
    >
      - This is a test
    </div>
    <button
      data-testid="Todo-delete-This_is_a_test"
    >
      x
    </button>
  </li>
</DocumentFragment>
`;
